## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [14]:
# View the column names in the data
df.columns
print(df.head())

       first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor  

In [20]:
# Use the describe function to gather some basic statistics
df.describe()


client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [24]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.info()



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [27]:
# What three item categories had the most entries?
category_counts = df['category'].value_counts()
top3_categories = category_counts.head(3)
print(top3_categories)

consumables    23538
furniture      11915
software        8400
Name: category, dtype: int64


In [28]:
# For the category with the most entries,
# which subcategory had the most entries?
filtered_df = df[df['category'] == 'consumables']
subcategory_most_entries = filtered_df['subcategory'].value_counts().idxmax()
subcategory_most_entries_count = filtered_df['subcategory'].value_counts().max()
print(subcategory_most_entries)
print(subcategory_most_entries_count)


bathroom supplies
6424


In [30]:
# Which five clients had the most entries in the data?

client_id_counts = df['client_id'].value_counts()
top5_clients = client_id_counts.head(5)
print(top5_clients)

33615    220
66037    211
46820    209
38378    207
24741    207
Name: client_id, dtype: int64


In [31]:
# Store the client ids of those top 5 clients in a list.
top5_clients_list = df['client_id'].value_counts().nlargest(5).index.tolist()
print(top5_clients_list)


[33615, 66037, 46820, 38378, 24741]


In [32]:
# How many total units (the qty column) did the
# client with the most entries order order?

filtered_df = df[df['client_id'] == 33615]
Total = filtered_df['qty'].sum()
print(Total)

64313


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [33]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty

df['line_subtotal'] = df['unit_price'] * df['qty']
#df
selected_columns = ['unit_price', 'qty', 'line_subtotal']
df_selected = df[selected_columns]
df_selected

unit_price   qty  line_subtotal
0         1096.80   105      115164.00
1           24.95    21         523.95
2           13.52    39         527.28
3           36.42    29        1056.18
4          195.10    20        3902.00
...           ...   ...            ...
54634       83.13    33        2743.29
54635      206.59    47        9709.73
54636       65.66   475       31188.50
54637        1.48   112         165.76
54638        3.01  1031        3103.31

[54639 rows x 3 columns]

In [34]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.
df['total_weight'] = df['unit_weight'] * df['qty']
def calculate_shipping_price(weight):
    if weight > 50:
        return 7 * weight
    else:
        return 10 * weight

df['shipping_price'] = df['total_weight'].apply(calculate_shipping_price)
sel_columns = ['unit_price', 'unit_weight', 'qty', 'total_weight', 'shipping_price']
df_sel = df[sel_columns]
df_sel


unit_price  unit_weight   qty  total_weight  shipping_price
0         1096.80         7.50   105        787.50         5512.50
1           24.95         1.49    21         31.29          312.90
2           13.52         1.68    39         65.52          458.64
3           36.42         1.23    29         35.67          356.70
4          195.10        46.43    20        928.60         6500.20
...           ...          ...   ...           ...             ...
54634       83.13         2.25    33         74.25          519.75
54635      206.59        11.70    47        549.90         3849.30
54636       65.66         4.16   475       1976.00        13832.00
54637        1.48        18.04   112       2020.48        14143.36
54638        3.01         2.07  1031       2134.17        14939.19

[54639 rows x 5 columns]

In [35]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%

df['line_price'] = ((df['line_subtotal'] + df['shipping_price'])*1.0925).round(2)

sel_columns = ['line_subtotal', 'shipping_price', 'line_price']
df_sel = df[sel_columns]
df_sel

line_subtotal  shipping_price  line_price
0          115164.00         5512.50   131839.08
1             523.95          312.90      914.26
2             527.28          458.64     1077.12
3            1056.18          356.70     1543.57
4            3902.00         6500.20    11364.40
...              ...             ...         ...
54634        2743.29          519.75     3564.87
54635        9709.73         3849.30    14813.24
54636       31188.50        13832.00    49184.90
54637         165.76        14143.36    15632.71
54638        3103.31        14939.19    19711.43

[54639 rows x 3 columns]

In [36]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).
df['line_cost'] = (df['unit_cost'] * df['qty']) + df['shipping_price']
df


first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ... unit_price unit_cost unit_weight   qty  \
0              17        2023  ...    1096.80    762.71        7.50   105   
1              20        2023  ...      24.95     15.09        1.49    21   
2               4        2023  ...      13.52      7.86        1.68    39   
3              17        2023  ...      36.42     24.85        1.23    29   
4              21        2023  ...     195.10    108.17       46.43    20   
...           ...         ...  ...        ...       ...         ...   ...   
54634          17        2023  ...      83.13     51.60        2.25    33   
54635           6        2023  ...     206.59    175.46       11.70    47   
54636           9        2023  ...      65.66     57.31        4.16   475   
54637           3        2023  ...       1.48      1.22       18.04   112   
54638           1        2023  ...       3.01      1.61        2.07  1031   

       line_number  line_subtotal  total_weight  shipping_price  line_price  \
0                1      115164.00        787.50         5512.50   131839.08   
1                0         523.95         31.29          312.90      914.26   
2                6         527.28         65.52          458.64     1077.12   
3                3        1056.18         35.67          356.70     1543.57   
4                1        3902.00        928.60         6500.20    11364.40   
...            ...            ...           ...             ...         ...   
54634            8        2743.29         74.25          519.75     3564.87   
54635            0        9709.73        549.90         3849.30    14813.24   
54636            6       31188.50       1976.00        13832.00    49184.90   
54637            0         165.76       2020.48        14143.36    15632.71   
54638            5        3103.31       2134.17        14939.19    19711.43   

       line_cost  
0       85597.05  
1         629.79  
2         765.18  
3        1077.35  
4        8663.60  
...          ...  
54634    2222.55  
54635   12095.92  
54636   41054.25  
54637   14280.00  
54638   16599.10  

[54639 rows x 23 columns]

In [37]:
# Create a column for the profit of
# each line using line cost and line price

df['line_profit'] = df['line_price'] - df['line_cost']
df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ... unit_cost unit_weight   qty  line_number  \
0              17        2023  ...    762.71        7.50   105            1   
1              20        2023  ...     15.09        1.49    21            0   
2               4        2023  ...      7.86        1.68    39            6   
3              17        2023  ...     24.85        1.23    29            3   
4              21        2023  ...    108.17       46.43    20            1   
...           ...         ...  ...       ...         ...   ...          ...   
54634          17        2023  ...     51.60        2.25    33            8   
54635           6        2023  ...    175.46       11.70    47            0   
54636           9        2023  ...     57.31        4.16   475            6   
54637           3        2023  ...      1.22       18.04   112            0   
54638           1        2023  ...      1.61        2.07  1031            5   

       line_subtotal  total_weight  shipping_price  line_price  line_cost  \
0          115164.00        787.50         5512.50   131839.08   85597.05   
1             523.95         31.29          312.90      914.26     629.79   
2             527.28         65.52          458.64     1077.12     765.18   
3            1056.18         35.67          356.70     1543.57    1077.35   
4            3902.00        928.60         6500.20    11364.40    8663.60   
...              ...           ...             ...         ...        ...   
54634        2743.29         74.25          519.75     3564.87    2222.55   
54635        9709.73        549.90         3849.30    14813.24   12095.92   
54636       31188.50       1976.00        13832.00    49184.90   41054.25   
54637         165.76       2020.48        14143.36    15632.71   14280.00   
54638        3103.31       2134.17        14939.19    19711.43   16599.10   

       line_profit  
0         46242.03  
1           284.47  
2           311.94  
3           466.22  
4          2700.80  
...            ...  
54634      1342.32  
54635      2717.32  
54636      8130.65  
54637      1352.71  
54638      3112.33  

[54639 rows x 24 colum

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [38]:
# Check your work using the totals above
order_ids_check = [2742071, 2173913, 6128929]
grouped_df = df.groupby('order_id').agg({'line_price': 'sum'}).reset_index()
filtered_df = grouped_df[grouped_df['order_id'].isin(order_ids_check)]
#filtered_df 
for index, row in filtered_df.iterrows():
    formatted_order_id = "{:.0f}".format(row['order_id'])
    formatted_price = "${:,.2f}".format(row['line_price'])
    print(f"Order ID {formatted_order_id} had a total price of {formatted_price}")

Order ID 2173913 had a total price of $162,388.71
Order ID 2742071 had a total price of $152,811.89
Order ID 6128929 had a total price of $923,441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [39]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.

top5_clients_list = df['client_id'].value_counts().nlargest(5).index.tolist()
grouped_df = df.groupby('client_id').agg({'line_price': 'sum'}).reset_index()
filtered_df = grouped_df[grouped_df['client_id'].isin(top5_clients_list)]
#filtered_df 
for index, row in filtered_df.iterrows():
    formatted_client_id = "{:.0f}".format(row['client_id'])
    formatted_price = "${:,.2f}".format(row['line_price'])
    print(f"{formatted_client_id}: {formatted_price}")

24741: $82,268,892.02
33615: $8,377,308.52
38378: $12,906,550.87
46820: $9,743,794.36
66037: $10,259,514.79


In [40]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.

top5_clients_list = df['client_id'].value_counts().nlargest(5).index.tolist()
grouped_df = df.groupby('client_id').agg({
    'qty': 'sum',
    'shipping_price': 'sum',
    'line_price': 'sum',
    'line_cost': 'sum',
    'line_profit': 'sum'
}).reset_index()
summary_df = grouped_df[grouped_df['client_id'].isin(top5_clients_list)]
summary_df

client_id     qty  shipping_price   line_price    line_cost  line_profit
144      24741  239862      5126448.37  82268892.02  45688899.71  36579992.31
258      33615   64313      1828984.89   8377308.52   6175313.91   2201994.61
319      38378   73667      3429455.40  12906550.87   9634720.98   3271829.89
402      46820   75768      1601448.84   9743794.36   7007191.64   2736602.72
621      66037   43018      1395151.85  10259514.79   7004482.98   3255031.81

In [41]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.
summary_df['line_price'] /= 1e6
summary_df['line_cost'] /= 1e6
summary_df['line_profit'] /= 1e6
summary_df['shipping_price'] /= 1e6

summary_df = summary_df.rename(columns=
                               {'line_price':'Total Revenue',
                                'line_cost':'Total Cost',
                                'line_profit':'Total Profit',
                                'qty': 'Units',
                                'client_id': 'Client ID',
                                'shipping_price': 'Shipping'
                               })
def format_currency(value):
    return "${:.2f} MM".format(value)

columns_to_format = ['Total Revenue', 'Total Cost', 'Total Profit', 'Shipping']
summary_df[columns_to_format] = summary_df[columns_to_format].applymap(format_currency)
summary_df


C:\Users\sebas\anaconda3\envs\dev\lib\site-packages\ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  after removing the cwd from sys.path.
C:\Users\sebas\anaconda3\envs\dev\lib\site-packages\ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """
C:\Users\sebas\anaconda3\envs\dev\lib\site-packages\ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

Client ID   Units  Shipping Total Revenue Total Cost Total Profit
144      24741  239862  $5.13 MM     $82.27 MM  $45.69 MM    $36.58 MM
258      33615   64313  $1.83 MM      $8.38 MM   $6.18 MM     $2.20 MM
319      38378   73667  $3.43 MM     $12.91 MM   $9.63 MM     $3.27 MM
402      46820   75768  $1.60 MM      $9.74 MM   $7.01 MM     $2.74 MM
621      66037   43018  $1.40 MM     $10.26 MM   $7.00 MM     $3.26 MM

In [42]:
# Sort the updated data by "Total Profit" form highest to lowest
summary_df = summary_df.sort_values(by='Total Profit', ascending=False)
summary_df

Client ID   Units  Shipping Total Revenue Total Cost Total Profit
144      24741  239862  $5.13 MM     $82.27 MM  $45.69 MM    $36.58 MM
319      38378   73667  $3.43 MM     $12.91 MM   $9.63 MM     $3.27 MM
621      66037   43018  $1.40 MM     $10.26 MM   $7.00 MM     $3.26 MM
402      46820   75768  $1.60 MM      $9.74 MM   $7.01 MM     $2.74 MM
258      33615   64313  $1.83 MM      $8.38 MM   $6.18 MM     $2.20 MM